Add content meta pattern

discodavey opened the pull request.

discodavey committed:
Adding styles to content-meta

discodavey committed:
Revert "Adding styles to content-meta"

This reverts commit 472885e62299d34a004a90970e6f05213e453f0c.

discodavey committed:
Revert "Revert "Adding styles to content-meta""

This reverts commit 1b954ee07873693505920ed818e7c734e37cd773.

discodavey committed:
Add mixins and colours for content-meta pattern

discodavey committed:
Amend colour from 888888 to 666666 to meet accessibility standards

discodavey committed:
remove as already in patterns

discodavey committed:
Add more styles for content-meta

davidcmoulton committed:
Merge branch 'master' into add-content-meta-pattern

davidcmoulton commented on src/shared-styles/_utilities.scss:
```diff
@@ -0,0 +1,10 @@
+@mixin inherit-all($ensure: ()) {
```
Our experience of using this approach has shown it's too blunt an instrument. Should remove this mixin and implement styling overrides as necessary on a case-by-case basis. This file can be deleted.

davidcmoulton commented on src/patterns/molecules/content-meta/content-meta.twig:
```diff
@@ -0,0 +1,38 @@
+
+  {%- endfor -%}
+
+</ul>
```
New line.

davidcmoulton commented on src/shared-styles/_spacing.scss:
```diff
@@ -0,0 +1,9 @@
+
+@mixin padding($sizes, $dimension: "") {
+  padding: 0;
+}
```
New line.

discodavey replied:
Not relevant as we have deleted this file

discodavey committed:
Add new way of font sizing

discodavey committed:
Merge remote-tracking branch 'upstream/master' into add-content-meta-pattern

davidcmoulton committed:
Merge branch 'master' into add-content-meta-pattern

davidcmoulton committed:
Use up-to-date sass modules

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -44,3 +64,10 @@
   line-height: $_baseline-size / $_font-size;
 }
+
+@mixin list-style-none() {
```
This doesn't address typography, so doesn't belong in this file. Would probably fit it above the style rules in `_normalize.scss`

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -35,6 +38,23 @@
+}
+
+@mixin label-tag-typography() {
+  @include _label-typography($color-primary-normal);
```
Colour needs namespacing.

davidcmoulton commented on src/patterns/molecules/content-meta/content-meta.scss:
```diff
@@ -0,0 +1,38 @@
+
+.content-meta {
+  @include typography.list-style-none();
+  @include spacing.block-spacing();
```
`@include space.block;`

davidcmoulton commented on src/patterns/molecules/content-meta/content-meta.scss:
```diff
@@ -0,0 +1,38 @@
+
+.content-meta__item {
+  display: inline-block;
+  @include typography.label-content-typography();
```
Reads awkwardly. ~Could improve by renaming the mixin to `label-content`, then this will read `@include typography.label-content()`, which is nicer.~ see https://github.com/libero/storybook/pull/31#discussion_r350086413

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -35,6 +38,23 @@
+  }
+}
+
+@mixin label-content-typography($color: color.$text-secondary, $uppercase: true) {
```
Shouldn't have `typography` in the name to avoid clumsy usage. Try renaming to `label-as-content`, so a usage would be `@include typography.label-as-content()`.

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -35,6 +38,23 @@
+  @include _label-typography($color, $uppercase);
+}
+
+@mixin label-tag-typography() {
```
Shouldn't have `typography` in the name to avoid clumsy usage. Perhaps rename it to `label-as-tag()`, then usage would read `@include typography.label-as-tag()`, which seems clearer.

davidcmoulton commented on src/shared-styles/_font.scss:
```diff
@@ -1,3 +1,4 @@
+$letterspacing-label: 0.5px !default;
```
Doesn't need to be configurable. Move to the `_label-typography` mixin as a private variable.

Should also rename to `$_letter-spacing` at the same time.

davidcmoulton reviewed (changes requested):
Needs tests for the new mixins, and the styles.

discodavey committed:
Code review changes

discodavey committed:
Remove as not needed

discodavey committed:
Remove centred in Storybook

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -35,6 +37,24 @@
+}
+
+@mixin label-as-tag() {
+  @include _label-typography(color.$color-primary-normal);
```
Colour name doesn't exist.

davidcmoulton replied:
Not addressed.

davidcmoulton commented on src/shared-styles/_normalize.scss:
```diff
@@ -71,3 +71,9 @@ strong {
 p {
   @include space.block();
 }
+
```
Move to before the style rules. Then if a style rule in this file uses it, things won't break.

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -35,6 +37,24 @@
 
+@mixin _label-typography($color, $uppercase: true) {
+  color: $color;
+  $letterspacing: 0.5px !default;
```
`$_letter-spacing`...

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -35,6 +37,24 @@
   @include _typography($scale-point, $baseline-units, font.$secondary, 600);
 }
 
+@mixin _label-typography($color, $uppercase: true) {
```
It's generally considered more readable to have private members below public ones, so  move this below `@mixin _typography`

davidcmoulton commented on src/patterns/molecules/content-meta/content-meta.stories.js:
```diff
@@ -0,0 +1,82 @@
+import './content-meta.scss';
+
+export default {
+  title: 'Molecules|ContentMeta',
```
`Content meta`

davidcmoulton reviewed (changes requested):
- This has inline-start padding:

<img width="421" alt="Screenshot 2019-11-25 15 10 29" src="https://user-images.githubusercontent.com/2893480/69552190-f13be700-0f95-11ea-88cf-63a02d5ac62c.png">

The Pattern Lab version of the ticket has no inline-start padding:

<img width="370" alt="Screenshot 2019-11-25 15 10 59" src="https://user-images.githubusercontent.com/2893480/69552143-dbc6bd00-0f95-11ea-8d21-1881e77d62e5.png">

- This is missing non-English language examples

- Still no Sass True tests...

discodavey committed:
Move to top of file

discodavey committed:
Change title

discodavey committed:
Add color

discodavey committed:
Move private to bottom of file

davidcmoulton commented:
I've discovered the cause of the padding not being set correctly, mentioned in https://github.com/libero/storybook/pull/31#pullrequestreview-322340251 above:  the`dir` attribute is missing from the `<html>` element generated by Storybook. The CSS selectors ultimately generated by [`logical.property()`](https://github.com/libero/storybook/blob/master/src/shared-styles/_logical.scss#L36) quite rightly depend on this `dir` value for the fallbacks, and also to enable the non-fallback property, where supported, to take precedence over the fallback. 

This is a wider issue than just this PR. Investigations show that Storybook doesn't set `dir` on `<html>` by default, and so none of the CSS selectors generated by `logical.property()` will match the desired elements within the Storybook interface.

This won't be a problem for using the patterns, as the implementation that incorporates them will need to set `dir` attributes on elements as appropriate anyway, but it is a problem for the build / test / demonstration of the patterns within Storybook.

Next steps: raising a ticket for the work to introduce a configurable `dir` attribute, and a configurable `lang` attribute onto the `<html>` element of Storybook pages. (There is a `lang="en"` attribute on the `<html>` element already, but it doesn't seem to be user configurable.)

Until that's in place, merging this PR is blocked.

discodavey committed:
Add different languages

discodavey committed:
Merge remote-tracking branch 'upstream/master' into add-content-meta-pattern

discodavey committed:
Add test for content-meta

davidcmoulton commented on src/patterns/molecules/content-meta/content-meta.stories.js:
```diff
@@ -0,0 +1,237 @@
+export const MetaAr = () => content(
+  {
+    attributes: {
+      dir: 'rtl',
```
See https://github.com/libero/storybook#setting-html-attributes-in-storybook on how to get the correct `dir` attribute set.

discodavey replied:
I got this from another pattern in storybook so if I need to change this then we need to do it across the whole library.

davidcmoulton replied:
Actually, no. It's fine as it is here. Leave it for now.

davidcmoulton commented on src/shared-styles/_color.scss:
```diff
@@ -1,2 +1,7 @@
+$primary-dark: rgb(2, 119, 189) !default;
+$text-normal: rgb(33, 33, 33) !default;
+$text-secondary: rgb(102, 102, 102) !default;
+$text-secondary__reverse: rgb(158, 158, 158) !default;
```
Looks like `$text-secondary` and `$primary-normal` are the only colours used in this PR. If the others aren't needed yet they shouldn't be included here.

discodavey replied:
primary-normal is being used as well.

discodavey replied:
$primary us being used as well.

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -35,6 +37,14 @@
   @include _typography($scale-point, $baseline-units, font.$secondary, 600);
 }
 
+@mixin label-as-content($color: color.$text-secondary, $uppercase: true) {
```
Still needs tests.

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -35,6 +37,14 @@
+  @include _label-typography($color, $uppercase);
+}
+
+@mixin label-as-tag() {
```
Still needs tests.

davidcmoulton replied:
This mixin isn't used (yet) and so should be removed.

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -44,3 +54,13 @@
+  color: $color;
+  $letterspacing: 0.5px !default;
+  @include _typography($scale-point: -3, $font-family: font.$secondary);
+  letter-spacing: $letterspacing;
```
~Move up to be beneath the line that sets the variable.~ No need if removing the variable.

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -44,3 +54,13 @@
+
+@mixin _label-typography($color, $uppercase: true) {
+  color: $color;
+  $letterspacing: 0.5px !default;
```
Remove the `!default` as we don't want this to be overridden. 

Rename to `$_letter-spacing` as it's local.

davidcmoulton replied:
Or even better, as discussed, just set it directly and lose the variable 😄

discodavey committed:
Code review changes

discodavey committed:
Remove variable

No need to declare variable and then call it in the next line as only used here.

discodavey committed:
Tests for label-as-content

discodavey committed:
Test for label-as-tag()

davidcmoulton commented on src/shared-styles/_typography.scss:
```diff
@@ -1,7 +1,9 @@
 @use "font";
 @use "rem";
 @use "scale";
+@use "space";
```
Not used?

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+
+@include describe('@mixin label-as-content') {
+
+  @include it('generates defaults for label-as-content') {
```
Don't need to say "for label-as-content" here, as this `it` clause is already inside the`describe('@mixin label-as-content') ` clause.

Suggest the slightly more descriptive `it('generates correct default styles')`.

davidcmoulton replied:
~Not done.~ Looking at old version, it's fine as now amended.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+      }
+
+      @include expect {
+        color: $text-secondary;
```
Adding local but exportable and overridable properties like this could be misleading and dangerous. Should either specify the color directly like this: `color: rgb(102, 102, 102)`, or use the variable from `_color.scss` like this: `color: color.$text-secondary`. As every other rule value in this test is a specified directly (e.g. the font family), then the colour should follow the same pattern and also be specified directly.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+
+  }
+
+  @include it('generates different color for text and default text-transform') {
```
Prefer `it('allows the color to be overridden')`

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+    @include assert() {
+
+      @include output {
+        @include typography.label-as-content($primary);
```
Set as an arbitrarily named local variable e.g. `$_new-color` just within this `assert` block.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -16,6 +16,11 @@
+$primary-normal: rgb(2, 136, 209) !default;
+$text-secondary: rgb(102, 102, 102) !default;
+
+
```
Remove. See comment relating to line 179.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+      }
+
+      @include expect {
+        color: $primary;
```
Use the local variable mentioned above.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+
+  }
+
+  @include it('generates default color and the text-transform is false') {
```
Prefer `it('allows the text-transform to be overridden')`

davidcmoulton replied:
Not done...

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+      }
+
+      @include expect {
+        color: $text-secondary;
```
Specify `color` directly.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+
+  }
+
+  @include it('generates different color and the text-transform is false') {
```
Prefer `it('allows both the color and the text-transform to be overridden')`

davidcmoulton replied:
This hasn't been done.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+    @include assert() {
+
+      @include output {
+        @include typography.label-as-content($color: $primary, $uppercase: false);
```
Use local variable for the colour here, scoped to the `assert`, as described in a comment for a previous test in this file.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+      }
+
+      @include expect {
+        color: $primary;
```
Use the local variable described above.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -159,3 +164,118 @@
+
+}
+
+@include describe('@mixin label-as-tag') {
```
The `label-as-tag` mixin that this tests for is not used, and so this `describe` block should be removed along with that mixin.

davidcmoulton commented on test/sass/normalize.spec.scss:
```diff
@@ -0,0 +1,34 @@
+
+        html[dir="ltr"] &:not([dir]),
+        &[dir="ltr"] {
+          padding-left: 0;
```
Need to suppress stylelint's logical property checking for this line to prevent unnecessary lint errors. See https://github.com/libero/storybook/blob/master/src/shared-styles/_normalize.scss#L59

davidcmoulton commented on test/sass/normalize.spec.scss:
```diff
@@ -0,0 +1,34 @@
+        }
+        html[dir="rtl"] &:not([dir]),
+        &[dir="rtl"] {
+          padding-right: 0;
```
See comment for line 21.

davidcmoulton commented on test/sass/normalize.spec.scss:
```diff
@@ -0,0 +1,34 @@
+@use "test" as *;
+@use "../../src/shared-styles/normalize";
+@use "../../src/shared-styles/space";
+@use "../../src/shared-styles/typography";
```
Not used.

discodavey committed:
Amendments from code review

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -16,9 +16,7 @@
 }
 
-$primary: rgb(2, 119, 189) !default;
 $primary-normal: rgb(2, 136, 209) !default;
```
Not needed.

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -189,16 +187,16 @@ $text-secondary: rgb(102, 102, 102) !default;
 
       @include output {
-        @include typography.label-as-content($primary);
+        @include typography.label-as-content($color: rgb(2, 119, 189));
```
See previous review https://github.com/libero/storybook/pull/31#discussion_r353845044

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -237,11 +235,11 @@ $text-secondary: rgb(102, 102, 102) !default;
 
       @include output {
-        @include typography.label-as-content($color: $primary, $uppercase: false);
+        @include typography.label-as-content($color: rgb(2, 119, 189), $uppercase: false);
```
As previously mentioned, use a local variable for colour here (as on line 197).

davidcmoulton reviewed (changes requested):
Almost there.

discodavey committed:
Amend tests from code review

davidcmoulton commented on test/sass/typograpy.spec.scss:
```diff
@@ -191,12 +188,14 @@ $primary-normal: rgb(2, 136, 209) !default;
 
     @include assert() {
 
+      $acolor: rgb(2, 119, 189);
```
Prefer to make this obviously local by prefixing with `_`, as mentioned. So `$_acolor`. Do that for all occurrences and I think this will be done.

discodavey committed:
Add underscore to acolor

davidcmoulton reviewed (changes requested):
One outstanding from previous review.

discodavey committed:
code review

davidcmoulton reviewed (changes requested):
Found another comment from a previous review that's not been addressed. Please can you go through and make sure that they're all done.

discodavey committed:
Amend code review and corrected spelling of file name

davidcmoulton commented on test/sass/typography.spec.scss:
```diff
@@ -159,3 +159,99 @@
+
+    @include assert() {
+
+      $_acolor: rgb(102, 102, 102);
```
Don't need this here. Line 222 should just specify the colour directly. As discussed, the colour only needs to be in a variable if it's used in both the `output` and the `expect` blocks. Here, it's only in the `expect` block so there's no duplication to remove, so there's no need a variable. (For an appropriate implementation using a variable see [within the assert block on line 236](https://github.com/libero/storybook/pull/31/files#diff-34e711f734d561d250826e618e148f3eR236).)

discodavey committed:
Code review
